pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,446 @@
+}
+
+// DeleteStream deletes the stream from the map
+func (s *FQDNDataServer) DeleteStream(stream standalonednsproxy.FQDNData_StreamPolicyStateServer) {
+	_, ok := s.streams.Load(stream)
+	if ok {
+		log.Infof("Deleting stream: %v", stream)
+		s.streams.Delete(stream)
+	} else {
+		log.Warnf("Stream not found: %v", stream)
+	}
+
+}
+
+// cleanupStreams handles the cleanup of streams when the server's context is cancelled.
+func (s *FQDNDataServer) cleanupStreams() {
+	s.streams.Range(func(key standalonednsproxy.FQDNData_StreamPolicyStateServer, cancelFunc context.CancelFunc) bool {
+		cancelFunc() // Ensure we cancel the context of each stream
+		s.streams.Delete(key)
+		return true
+	})
+	log.Info("All streams have been cleaned up")
+}
+
+// UpdateMappingRequest updates the FQDN mapping with the given data
+// SDP sends the fqdn mapping to cilium agent
+// Steps to update the mapping:
+// 1. Get the endpoint from the IP
+// 2. If the endpoint is not found, return an error
+// 3. If the IPs are not empty, update the cilium agent with the mapping
+func (s *FQDNDataServer) UpdateMappingRequest(ctx context.Context, mappings *standalonednsproxy.FQDNMapping) (*standalonednsproxy.UpdateMappingResponse, error) {
+	log.Debugf("UpdateMappings %v", mappings)
+	now := time.Now()
+	var ips []netip.Addr
+
+	endpointAddr := netip.MustParseAddr(string(mappings.SourceIp))
+
+	ep := s.endpointManager.LookupIP(endpointAddr)
+	if ep == nil {
+		log.Errorf("endpoint not found for IP: %s", mappings.SourceIp)
+		return &standalonednsproxy.UpdateMappingResponse{}, fmt.Errorf("endpoint not found for IP: %s", mappings.SourceIp)
+	}
+
+	recordIps := mappings.GetRecordIp()
+	if len(recordIps) == 0 {
+		// We don't have any IPs to update the mappings with
+		return &standalonednsproxy.UpdateMappingResponse{
+			Response: standalonednsproxy.ResponseCode_RESPONSE_CODE_NO_ERROR,
+		}, nil
+	}
+
+	for _, ip := range recordIps {
+		ips = append(ips, netip.MustParseAddr(string(ip)))
+	}
+
+	if mappings.GetResponseCode() == dns.RcodeSuccess {
+		err := s.updateOnDNSMsg(now, ep, mappings.GetFqdn(), ips, int(mappings.GetTtl()), nil)
+		if err != nil {
+			return &standalonednsproxy.UpdateMappingResponse{}, fmt.Errorf("cannot update DNS cache: %w", err)
+		}
+	}
+
+	return &standalonednsproxy.UpdateMappingResponse{
+		Response: standalonednsproxy.ResponseCode_RESPONSE_CODE_NO_ERROR,
+	}, nil
+}
+
+// RunServer starts the Standalone DNS Proxy grpc server on the given port
+func RunServer(port int, server *FQDNDataServer) error {
+	address := fmt.Sprintf("localhost:%d", port)
+	log.Infof("Starting Standalone DNS Proxy grpc service on %s", address)
+	lis, err := net.Listen("tcp", address)
+	if err != nil {
+		log.Errorf("failed to listen: %v", err)
+		return err
+	}
+	grpcServer := grpc.NewServer(grpc.KeepaliveEnforcementPolicy(kaep), grpc.KeepaliveParams(kasp))
+	standalonednsproxy.RegisterFQDNDataServer(grpcServer, server)
+
+	if err := grpcServer.Serve(lis); err != nil {
```

Comment:
Isn't `Serve` blocking? As in, if it returns the server is already dead and it hence makes no sense to setup the `closer` func synchronously after?